misk/src/main/kotlin/misk/digester/MergingDigest.kt

```diff
@@ -0,0 +1,351 @@
+        mergedWeight += nextMain.weight
+      } else {
+        // before merging, we have to save the next main centroid somewhere
+        // else, so that we don't overwrite it
+        if (actualMainCentroids.size != 0) {
+          swappedCentroids.add(actualMainCentroids[0])
+          actualMainCentroids.removeAt(0)
+        }
+        tempIndex++
+
+        lastMergedIndex = mergeOne(mergedWeight, totalWeight, lastMergedIndex, nextTemp)
+        mergedWeight += nextTemp.weight
+      }
+    }
+
+    tempCentroids.clear()
+    tempWeight = 0.0
+    mainWeight = totalWeight
+  }
+
+  /**
+   * merges a single centroid into the mergedCentroids list
+   * note that "merging" sometimes creates a new centroid in the list, however
+   * the length of the list has a strict upper bound (see constructor)
+   */
+  fun mergeOne(
+      beforeWeight: Double,
+      totalWeight: Double,
+      beforeIndex: Double,
+      next: Centroid
+  ): Double {
+
+    // compute the quantile index of the element we're about to merge
+    var nextIndex = indexEstimate((beforeWeight + next.weight) / totalWeight)
+
+    if (nextIndex-beforeIndex > 1 || mainCentroids.size == 0) {
+      // the new index is far away from the last index of the current centroid
+      // therefore we cannot merge into the current centroid or it would
+      // become too wide, so we will append a new centroid
+      mainCentroids.add(next)
+      // return the last index that was merged into the previous centroid
+      return indexEstimate(beforeWeight / totalWeight)
+    } else {
+      // the new index fits into the range of the current centroid, so we
+      // combine it into the current centroid's values
+      // this computation is known as welford's method, the order matters
+      // weight must be updated before mean
+
+      //TODO: is this thread safe? doublecheck that this successfully updates all properties
+      var mainCentroidsTemp = mainCentroids[mainCentroids.size-1]
+      var builder = mainCentroidsTemp.newBuilder()
+      builder.weight += next.weight
+      builder.mean +=
+          (next.mean - mainCentroids[mainCentroids.size-1].mean) * next.weight / mainCentroids[mainCentroids.size-1].weight
+
+      // we did not create a new centroid, so the trailing index of the previous
+      // centroid remains
+      return beforeIndex
+    }
+  }
+
+  /**
+   * given a quantile, estimate the index of the centroid that contains it using
+   * the given compression
+   */
+  fun indexEstimate(quantile: Double) : Double {
+    return compression * ((asin(2*quantile-1) / PI) + 0.5).toFloat()
+  }
+
+  /**
+   * Returns a value such that the fraction of values in td below that value is
+   * approximately equal to quantile. Returns NaN if the digest is empty.
+   */
+  fun quantile(quantile: Double): Double {
+    if (quantile < 0 || quantile > 1) {
+      throw IndexOutOfBoundsException("quantile out of bounds")
```

Comment:
`IllegalArgumentException`. `IndexOutOfBoundsException` is used for lists/arrays.

Reply:
```kotlin
require(quantile >= 0 && quantile <= 1) { "quantile must be between 0 and 1" }
```